hfile_s3.c

```diff
@@ -1113,330 +1074,1281 @@ static int write_authorisation_callback(void *auth, char *request, kstring_t *co
+
+        if (fp->buffer.l && fp->last_read_buffer < fp->buffer.l) {
+            // copy data across
+            size_t to_copy;
+            size_t remaining = fp->buffer.l - fp->last_read_buffer;
+            size_t bytes_left = nbytes - read;
+
+            if (hts_verbose >  5) fprintf(stderr, "remaining %zu read %zu bytes_left %zu, nbytes %zu\n", remaining, read, bytes_left, nbytes);
+
+            if (bytes_left < remaining) {
+                to_copy = bytes_left;
+            } else {
+                to_copy = remaining;
+            }
+
+            memcpy(buffer + read, fp->buffer.s + fp->last_read_buffer, to_copy);
+            read += to_copy;
+            fp->last_read_buffer += to_copy;
+
+            if ((fp->buffer.l < fp->part_size) && (fp->last_read_buffer == fp->buffer.l)) {
+                fp->keep_going = 0;
+            }
+        } else {
+            int ret;
+            kstring_t response = {0, 0, NULL};
+
+            ret = get_part(fp, &response);
+
+            if (hts_verbose > 5) fprintf(stderr, "read error %d\n", ret);
+
+            if (ret < 0)
+                return ret;
+
+            if (fp->buffer.l == 0) {
+                fprintf(stderr, "Returned no data.\n");
+                fp->keep_going = 0;
+                break;
+            }
+
+            fp->last_read_buffer = 0;
+            fp->last_read = fp->last_read + fp->buffer.l;
+            gets++;
+
+            // deal with the response
+            ks_free(&response);
+        }
+    }
+
+    total += read;
+    if (hts_verbose > 5) fprintf(stderr, "s3_read end read  %zu nbytes %zu total %zu gets %zu\n", read, nbytes, total, gets);
+
+    return read;
+}
+
+
+static off_t s3_seek(hFILE *fpv, off_t offset, int whence) {
+    hFILE_s3 *fp = (hFILE_s3 *)fpv;
+    off_t origin;
+
+    if (fp->write) {
+        // lets not try and seek while writing
+        errno = ESPIPE;
+        return -1;
+    }
+
+    // I am not sure we handle any seek other than one from the beginning
+    switch (whence) {
+        case SEEK_SET:
+            origin = 0;
+            break;
+        case SEEK_CUR:
+            // we might be able to work this out
+            errno = ENOSYS;
+            return -1;
+        case SEEK_END:
+            if (fp->file_size < 0) {
+                errno = ESPIPE;
+                return -1;
+            }
+
+            origin = fp->file_size;
+            break;
+        default:
+            errno = EINVAL;
+            return -1;
+    }
+
+    // Check 0 <= origin+offset < fp->file_size carefully, avoiding overflow
+    if ((offset < 0)? origin + offset < 0
+                : (fp->file_size >= 0 && offset > fp->file_size - origin)) {
+        errno = EINVAL;
+        return -1;
+    }
+
+    fp->keep_going = 1;
+    fp->last_read = origin + offset; // origin is really only useful if we can make the other modes work
+    ks_clear(&fp->buffer); // resetting fp->buffer triggers a new remote read
```

Comment:
It would be nice to work out if the destination is within the range of the existing buffer, and reuse it if that's the case.  It might allow some upstream requests to be avoided.